pkg/framework/plugins/removefailedpods/failedpods.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package removefailedpods
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/descheduler/pkg/apis/componentconfig"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	podutil "sigs.k8s.io/descheduler/pkg/descheduler/pod"
+	"sigs.k8s.io/descheduler/pkg/framework"
+)
+
+const PluginName = "RemoveFailedPods"
+
+// RemoveFailedPods evicts pods on the node which violate NoSchedule Taints on nodes
+type RemoveFailedPods struct {
+	handle    framework.Handle
+	args      *componentconfig.RemoveFailedPodsArgs
+	podFilter podutil.FilterFunc
+}
+
+var _ framework.Plugin = &RemoveFailedPods{}
+var _ framework.DeschedulePlugin = &RemoveFailedPods{}
+
+// New builds plugin from its arguments while passing a handle
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	failedPodsArgs, ok := args.(*componentconfig.RemoveFailedPodsArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type RemoveFailedPodsArgs, got %T", args)
+	}
+
+	var includedNamespaces, excludedNamespaces sets.String
+	if failedPodsArgs.Namespaces != nil {
+		includedNamespaces = sets.NewString(failedPodsArgs.Namespaces.Include...)
+		excludedNamespaces = sets.NewString(failedPodsArgs.Namespaces.Exclude...)
+	}
+
+	podFilter, err := podutil.NewOptions().
+		WithFilter(handle.Evictor().Filter).
+		WithNamespaces(includedNamespaces).
+		WithoutNamespaces(excludedNamespaces).
+		WithLabelSelector(failedPodsArgs.LabelSelector).
+		BuildFilterFunc()
+	if err != nil {
+		return nil, fmt.Errorf("error initializing pod filter function: %v", err)
+	}
+
+	podFilter = podutil.WrapFilterFuncs(func(pod *v1.Pod) bool { return pod.Status.Phase == v1.PodFailed }, podFilter)
+
+	if len(failedPodsArgs.ExcludeOwnerKinds) > 0 {
+		excludeOwnerKinds := sets.NewString(failedPodsArgs.ExcludeOwnerKinds...)
+		podFilter = podutil.WrapFilterFuncs(podFilter, func(pod *v1.Pod) bool {
+			ownerRefList := podutil.OwnerRef(pod)
+			for _, owner := range ownerRefList {
+				if excludeOwnerKinds.Has(owner.Kind) {
```

Comment:
It is still worth logging the exclusion. E.g. `klog.V(4).InfoS("owner kind is excluded, excluding pod", "pod", klog.KObj(pod), "kind", owner.Kind)`. The same for MinPodLifetimeSeconds and Reasons.

Reply:
Makes sense. I ended up bringing the current logic back since it writes the errors as an aggregate error.